### PR TITLE
fix(ivy): fix type error in newer version of TS

### LINKED ISSUE #22843: [6.0.0-beta.8] throws error in @angular/core/src/render3/interfaces/definition.d.ts
## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[x] Bug report  
[ ] Feature request
[ ] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
</code></pre>

## Current behavior

I just upgraded from `@angular/cli` `6.0.0-beta.5` to `6.0.0-beta.6` which also pulled `@angular/angular` `6.0.0-beta.8` (didn't know it existed, because it's not listed here on github).

Though I'm now using `typescript 2.7.2` with `strict: true`.
And `ng serve --aot=true` with `es2015`.

Here's the error I get at start up:
```
ERROR in node_modules/@angular/core/src/render3/interfaces/definition.d.ts(250,18): error TS2430: Interface 'ComponentDefArgs<T>' incorrectly extends interface 'DirectiveDefArgs<T>'.
  Types of property 'features' are incompatible.
    Type 'ComponentDefFeature[] | undefined' is not assignable to type 'DirectiveDefFeature[] | undefined'.
      Type 'ComponentDefFeature[]' is not assignable to type 'DirectiveDefFeature[] | undefined'.
        Type 'ComponentDefFeature[]' is not assignable to type 'DirectiveDefFeature[]'.
          Type 'ComponentDefFeature' is not assignable to type 'DirectiveDefFeature'.
            Types of parameters 'componentDef' and 'directiveDef' are incompatible.
              Type 'DirectiveDef<T>' is not assignable to type 'ComponentDef<T>'.
                Property 'tag' is missing in type 'DirectiveDef<T>'.
```

## Expected behavior
No error.

## Minimal reproduction of the problem with instructions

I'm not sure why I'm getting this error. I've not enabled Ivy.
And I'm getting **no** error when creating a new minimal project using `ng new projectName && cd projectName && ng update --next`.

## What is the motivation / use case for changing the behavior?

Cannot start my application at all. That's sad :(

## Environment

<pre><code>
Angular version: 6.0.0-beta.8
TypeScript: 2.7.2 (with `strict: true`)

Browser:
- [x] Chrome (desktop) version 65
- [ ] Chrome (Android) version XX
- [ ] Chrome (iOS) version XX
- [ ] Firefox version XX
- [ ] Safari (desktop) version XX
- [ ] Safari (iOS) version XX
- [ ] IE version XX
- [ ] Edge version XX
 
For Tooling issues:
- Node version: 9.7.1
- Platform:  macOS 10.13.3

Others:

</code></pre>

### LINKED ISSUE #22877: compiling aio against 6.0.0-beta.8 fails with Interface 'ComponentDefArgs<T>' incorrectly extends interface 'DirectiveDefArgs<T>'
```
ERROR in node_modules/@angular/core/src/render3/interfaces/definition.d.ts(250,18): error TS2430: Interface 'ComponentDefArgs<T>' incorrectly extends interface 'DirectiveDefArgs<T>'.
  Types of property 'features' are incompatible.
    Type 'ComponentDefFeature[] | undefined' is not assignable to type 'DirectiveDefFeature[] | undefined'.
      Type 'ComponentDefFeature[]' is not assignable to type 'DirectiveDefFeature[] | undefined'.
        Type 'ComponentDefFeature[]' is not assignable to type 'DirectiveDefFeature[]'.
          Type 'ComponentDefFeature' is not assignable to type 'DirectiveDefFeature'.
            Types of parameters 'componentDef' and 'directiveDef' are incompatible.
              Type 'DirectiveDef<T>' is not assignable to type 'ComponentDef<T>'.
                Property 'tag' is missing in type 'DirectiveDef<T>'.
```

env info:

```
Angular: 6.0.0-beta.8
... animations, common, compiler, compiler-cli, core, elements
... forms, http, platform-browser, platform-browser-dynamic
... platform-server, router

@angular/cdk: 5.0.0-rc.1
@angular/cli: 1.7.3
@angular/material: 5.0.0-rc.1
@angular/service-worker: 1.0.0-beta.16
@angular-devkit/build-optimizer: 0.3.2
@angular-devkit/core: 0.3.2
@angular-devkit/schematics: 0.3.2
@ngtools/json-schema: 1.2.0
@ngtools/webpack: 1.10.2
@schematics/angular: 0.3.2
@schematics/package-update: 0.3.2
typescript: 2.7.2
webpack: 3.11.0

```

### PR DESCRIPTION
**Pull request by mhevery**
Newer version of TS is stricter about types and flags counter-variant
types in some  situations. This change inlines the DirectiveDefArgs
into the arguments which:
1) removes the inheritance which caused the issue and
2) Makes it more friendly to IDEs since they will not report comments.

Closes #22877
Closes #22843

## PR Checklist
Please check if your PR fulfills the following requirements:

- [ ] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] angular.io application / infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?

Issue Number: N/A

## Does this PR introduce a breaking change?
```
[ ] Yes
[ ] No
```

### CODE REVIEW

File: packages/core/src/render3/definition.ts [hunk 1]
```diff
@@ -34,14 +35,126 @@ import {ComponentDef, ComponentDefArgs, DirectiveDef, DirectiveDefArgs, Directiv
...
+   * - `elementClass`
+   * - `elementStyle`
+   *
+   */
+  template: ComponentTemplate<T>;
+
+  /**
+   * A list of optional features to apply.
+   *
+   * See: {@link NgOnChancesFeature}, {@link PublicFeature}
```
Comment (kara) [inline packages/core/src/render3/definition.ts]: `NgOnChangesFeature`

File: packages/core/src/render3/definition.ts [hunk 2]
```diff
@@ -34,14 +35,126 @@ import {ComponentDef, ComponentDefArgs, DirectiveDef, DirectiveDefArgs, Directiv
...
+}): ComponentDef<T> {
   const type = componentDefinition.type;
   const def = <ComponentDef<any>>{
     type: type,
     diPublic: null,
     factory: componentDefinition.factory,
-    tag: (componentDefinition as ComponentDefArgs<T>).tag || null !,
-    template: (componentDefinition as ComponentDefArgs<T>).template || null !,
+    tag: componentDefinition.tag || null !,
+    template: componentDefinition.template || null !,
```
Comment (kara) [inline packages/core/src/render3/definition.ts]: Since `tag` and `template` are required, do we still need the `|| null` for them here?

**Review by kara — APPROVED** [review]
Nits

File: packages/core/src/render3/definition.ts (hunk 2 above)
Comment (kara) [reply packages/core/src/render3/definition.ts]: Nevermind, I see that it's re-used for `defineDirective`.

File: packages/core/src/render3/definition.ts (hunk 1 above)
Comment (IgorMinar) [reply packages/core/src/render3/definition.ts]: @mhevery plz fix

Comment (mhevery) [reply packages/core/src/render3/definition.ts]: fixed

**Pull request closed without merging**